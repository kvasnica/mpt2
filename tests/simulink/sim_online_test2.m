function sim_regul_test1

% tests simulation of on-line controllers, tracking problems
x0 = [1; 1];
assignin('base', 'sim_x0', x0);

%----------------------------------------------------------
% nx == ny
%----------------------------------------------------------
ref = [-1; 0];
assignin('base', 'sim_ref', ref);
Double_Integrator
assignin('base', 'sim_sysStruct', sysStruct);


%----------------------------------------------------------
% tracking=1
clear sysStruct probStruct
Double_Integrator
probStruct.N = 2;
probStruct.tracking = 1;
ctrl = mpt_control(sysStruct, probStruct, 'online');
assignin('base', 'sim_ctrl', ctrl);
sim('sim_online');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to reference?
mbg_asserttrue(norm(sim_x(end, :)' - ref) < 0.11);



%----------------------------------------------------------
% tracking=2
clear sysStruct probStruct
Double_Integrator
probStruct.N = 2;
probStruct.tracking = 2;
ctrl = mpt_control(sysStruct, probStruct, 'online');
assignin('base', 'sim_ctrl', ctrl);
sim('sim_online');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to reference?
mbg_asserttrue(norm(sim_x(end, :)' - ref) < 0.06);



%----------------------------------------------------------
% tracking=1, mpt_ownmpc()
clear sysStruct probStruct
Double_Integrator
probStruct.N = 2;
probStruct.tracking = 1;
[C, O, V] = mpt_ownmpc(sysStruct, probStruct, 'online');
ctrl = mpt_ownmpc(sysStruct, probStruct, C, O, V, 'online');
assignin('base', 'sim_ctrl', ctrl);
sim('sim_online');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to reference?
mbg_asserttrue(norm(sim_x(end, :)' - ref) < 0.11);



%----------------------------------------------------------
% tracking=2, mpt_ownmpc()
clear sysStruct probStruct
Double_Integrator
probStruct.N = 2;
probStruct.tracking = 2;
[C, O, V] = mpt_ownmpc(sysStruct, probStruct, 'online');
ctrl = mpt_ownmpc(sysStruct, probStruct, C, O, V, 'online');
assignin('base', 'sim_ctrl', ctrl);
sim('sim_online');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to reference?
mbg_asserttrue(norm(sim_x(end, :)' - ref) < 0.06);



%----------------------------------------------------------
% ny ~= nx
%----------------------------------------------------------
Double_Integrator
sysStruct.C = [1 0]; sysStruct.D = 0; sysStruct.ymax=5; sysStruct.ymin=-5;
assignin('base', 'sim_sysStruct', sysStruct);
ref = -1;
assignin('base', 'sim_ref', ref);


%----------------------------------------------------------
% tracking=1, ny~=nx
clear sysStruct probStruct
Double_Integrator
sysStruct.C = [1 0]; sysStruct.D = 0; sysStruct.ymax=5; sysStruct.ymin=-5;
probStruct.Qy = 10;
probStruct.N = 2;
probStruct.tracking = 1;
ctrl = mpt_control(sysStruct, probStruct, 'online');
assignin('base', 'sim_ctrl', ctrl);
sim('sim_online');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to reference?
mbg_asserttrue(norm(sim_y(end, :)' - ref) < 5e-4);



%----------------------------------------------------------
% tracking=2, ny~=nx
clear sysStruct probStruct
Double_Integrator
sysStruct.C = [1 0]; sysStruct.D = 0; sysStruct.ymax=5; sysStruct.ymin=-5;
probStruct.Qy = 10;
probStruct.N = 2;
probStruct.tracking = 2;
ctrl = mpt_control(sysStruct, probStruct, 'online');
assignin('base', 'sim_ctrl', ctrl);
sim('sim_online');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to reference?
mbg_asserttrue(norm(sim_y(end, :)' - ref) < 1.3e-3);



%----------------------------------------------------------
% tracking=1, ny~=nx, mpt_ownmpc()
clear sysStruct probStruct
Double_Integrator
sysStruct.C = [1 0]; sysStruct.D = 0; sysStruct.ymax=5; sysStruct.ymin=-5;
probStruct.Qy = 10;
probStruct.N = 2;
probStruct.tracking = 1;
[C, O, V] = mpt_ownmpc(sysStruct, probStruct, 'online');
ctrl = mpt_ownmpc(sysStruct, probStruct, C, O, V, 'online');
assignin('base', 'sim_ctrl', ctrl);
sim('sim_online');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to reference?
mbg_asserttrue(norm(sim_y(end, :)' - ref) < 5e-4);



%----------------------------------------------------------
% tracking=2, ny~=nx, mpt_ownmpc()
clear sysStruct probStruct
Double_Integrator
sysStruct.C = [1 0]; sysStruct.D = 0; sysStruct.ymax=5; sysStruct.ymin=-5;
probStruct.Qy = 10;
probStruct.N = 2;
probStruct.tracking = 2;
[C, O, V] = mpt_ownmpc(sysStruct, probStruct, 'online');
ctrl = mpt_ownmpc(sysStruct, probStruct, C, O, V, 'online');
assignin('base', 'sim_ctrl', ctrl);
sim('sim_online');

% solution always feasible?
mbg_asserttrue(all(sim_reg ~= 0));
% solution converged to reference?
mbg_asserttrue(norm(sim_y(end, :)' - ref) < 1.3e-3);

