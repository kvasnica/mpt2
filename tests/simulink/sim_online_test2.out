Using an LTI model...
WARNING: no state constraints defined, problem can be badly scaled...

The controller will return deltaU=u(k)-u(k-1) instead of u(k) when evaluated!
================================================================================
WARNING: offset-free tracking cannot be guaranteed with probStruct.tracking=2 !
================================================================================

Using an LTI model...
WARNING: no state constraints defined, problem can be badly scaled...

State vector will have the format x = [x0; u(k-1); reference].
Using an LTI model...
WARNING: no state constraints defined, problem can be badly scaled...

================================================================================
WARNING: offset-free tracking cannot be guaranteed with probStruct.tracking=2 !
================================================================================

State vector will have the format x = [x0; reference].
Using an LTI model...
WARNING: no state constraints defined, problem can be badly scaled...

WARNING: Closed-loop stability not guaranteed for penalties on outputs.
Using an LTI model...
WARNING: no state constraints defined, problem can be badly scaled...

The controller will return deltaU=u(k)-u(k-1) instead of u(k) when evaluated!
WARNING: Closed-loop stability not guaranteed for penalties on outputs.
================================================================================
WARNING: offset-free tracking cannot be guaranteed with probStruct.tracking=2 !
================================================================================

Using an LTI model...
WARNING: no state constraints defined, problem can be badly scaled...

State vector will have the format x = [x0; u(k-1); reference].
Using an LTI model...
WARNING: no state constraints defined, problem can be badly scaled...
WARNING: Closed-loop stability not guaranteed for penalties on outputs.

================================================================================
WARNING: offset-free tracking cannot be guaranteed with probStruct.tracking=2 !
================================================================================

State vector will have the format x = [x0; reference].
Using an LTI model...
WARNING: no state constraints defined, problem can be badly scaled...
WARNING: Closed-loop stability not guaranteed for penalties on outputs.
